### DISCUSSION TITLE
How to release memory when reading images from buffers

### DISCUSSION THREAD

**Discussion by enricoribelli** [davidbyttow/govips#406] [Q&A]
I am trying to build an API that can accept many http calls per second. For each of these, my service will stream down into memory an image taken from a URL, do some modifications on top of it and then serve it back (or currently just saving the image file).
I am trying to load test this service and I am noticing that the memory doesn't get released from GoVips and I am wondering if I am doing something wrong.

```
// Grabbing the image:
resp, err := http.DefaultClient.Do(req)
	
defer resp.Body.Close()

body, err := io.ReadAll(resp.Body)
	

image := new(vips.ImageRef)
image, err = vips.LoadImageFromBuffer(body, nil)
    


// some resizing logic with some params calculated somewhere else
err := image.Resize(scale, vips.KernelLanczos3)
	
		
// and some extra manipulation
err := image.ToColorSpace(vips.InterpretationSRGB)

// streaming out
imagebytes, _, err := image.ExportJpeg(ep)
ep := vips.NewJpegExportParams()
ep.StripMetadata = true
ep.Quality = 92
ep.Interlace = false
err = os.WriteFile(fmt.Sprintf("./output_%s.jpg", file_format), imagebytes, 0644)
```
am I doing something visibly wrong or should I correctly expect the memory to be released properly by itself once the image is saved? The problem here is that most likely there is a go routine being spawned from every http call reaching the API.. I just don't know what happens next when it comes to GoVips

**Comment by tonimelisma** [comment]
Garbage collection is difficult in golang. govips can be pretty efficient and gobble more memory than the gc can release.

https://tip.golang.org/doc/gc-guide

You can adjust some gc parameters to make it more aggressive:

https://pkg.go.dev/runtime

You could also try putting the image processing parts into goroutines or even separate processes.

**Comment by n0vad3v** [comment]
> my service will stream down into memory an image taken from a URL, do some modifications on top of it and then serve it back

Sounds like your service is quite like us https://github.com/webp-sh/webp_server_go and https://webp.se 😃

Maybe you could try using `img.Close()` after usage? And the following changes might be helpful:
* Change `MALLOC_ARENA_MAX` to 1 
* Change to another malloc other than the default glibc, we have a comparison here https://docs.webp.sh/benchmark/ and related discussion here: https://github.com/webp-sh/webp_server_go/issues/198
